### PR TITLE
Added Unit tests

### PR DESCRIPTION
**Pull request by topguns837**
- [x] Made the line_follower and naive_line_follower compile
- [x] Added basic unit tests for both line_follower and naive_line_follower
- [x]  Added good unit tests for line_follower

### CODE REVIEW

File: src/naive_line_follower.cpp [hunk 1]
```diff
@@ -1,86 +1,116 @@
...
+    waypoints = waypoints_param.as_double_array();
+    waypoint_1_.x = waypoints[0]; waypoint_1_.y = waypoints[1]; waypoint_1_.z = waypoints[2];
+    waypoint_2_.x = waypoints[3]; waypoint_2_.y = waypoints[4]; waypoint_2_.z = waypoints[5];
+    waypoint_angle = std::atan2(waypoint_2_.y  - waypoint_1_.y , waypoint_2_.x  - waypoint_1_.x );
 
-  void goToGoal() {
-    // This function gets called at a fixed time interval to update the robot's movement.
-    // For this example, we apply a simple proportional control based on the robot's current position
-    // and the line formed by the waypoints.
-    geometry_msgs::msg::TransformStamped transformStamped;
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp:49]: Use underscores for variable names all the way through

File: src/naive_line_follower.cpp [hunk 2]
```diff
@@ -1,86 +1,116 @@
...
+
+    if(!yaw_fixed){
+
+      // Convert Quaternion to Euler
+      tf2::Quaternion q(transformStamped.transform.rotation.x, transformStamped.transform.rotation.y, transformStamped.transform.rotation.z, transformStamped.transform.rotation.w);
+      tf2::Matrix3x3 m(q);
+      m.getRPY(bot_roll, bot_pitch, bot_yaw);
+
+      // Convert the bot yaw angle from range[-pi, pi] to [0, 2*pi]
+      if (bot_yaw<=0){
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp]: Should be <, not <=

File: src/naive_line_follower.cpp [hunk 3]
```diff
@@ -1,86 +1,116 @@
...
+    // Lookup transform to get the bot's position
+    transformStamped = tf_buffer_.lookupTransform("empty_world", "bcr_bot", tf2::TimePointZero, std::chrono::milliseconds(100));
+
+    if(!yaw_fixed){
+
+      // Convert Quaternion to Euler
+      tf2::Quaternion q(transformStamped.transform.rotation.x, transformStamped.transform.rotation.y, transformStamped.transform.rotation.z, transformStamped.transform.rotation.w);
+      tf2::Matrix3x3 m(q);
+      m.getRPY(bot_roll, bot_pitch, bot_yaw);
+
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp]: We don't need the initial alignment, we make the assumption that robot is sufficiently well aligned to begin with.

File: src/naive_line_follower.cpp [hunk 4]
```diff
@@ -1,86 +1,116 @@
...
+        goal_x * transformStamped.transform.translation.y +
+        waypoint_2_.x * waypoint_1_.y -
+        waypoint_2_.y * waypoint_1_.x) /
+        std::sqrt(std::pow(goal_y, 2) + std::pow(goal_x, 2));
 
-      // Populate cmd_vel
-      cmd_vel_.linear.x = 0.5; // constant forward velocity
-      cmd_vel_.angular.z = -control_effort; // turning rate
+      // Apply proportional control
+      double k_p = 0.1;
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp]: Initialize in constructor. Make a function called setControlGain to change this if needed.

File: src/naive_line_follower.cpp [hunk 5]
```diff
@@ -1,86 +1,116 @@
...
 
-      // Publish cmd_vel
+      // Calculate the distance effort from the distance error
+      distance_effort = k_p * distance_error;
+      
+      // Create and publish the Twist message
+      geometry_msgs::msg::Twist cmd_vel;
+      cmd_vel_.linear.x = 1.0; // constant forward velocity
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp]: Move this to the constructor

File: src/naive_line_follower.cpp [hunk 6]
```diff
@@ -1,86 +1,116 @@
...
-  tf2_ros::Buffer tf_buffer_;
-  tf2_ros::TransformListener tf_listener_;
-  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
-  rclcpp::TimerBase::SharedPtr control_loop_timer_;
-  geometry_msgs::msg::Point waypoint_1_;
-  geometry_msgs::msg::Point waypoint_2_;
-  geometry_msgs::msg::Twist cmd_vel_;
+  // Create a timer to call goToGoal
+  control_loop_timer_ = this->create_wall_timer(
+    std::chrono::milliseconds(100),
```
Comment (ggupta9777) [inline src/naive_line_follower.cpp]: Control loop should faster, say at least 20 Hz

File: src/line_follower.cpp [hunk 7]
```diff
@@ -1,77 +1,144 @@
...
 
-  // Function to calculate the control effort from the error.
-  double calculateControlEffort(double error) {
-    double k_p = 1.0; // Proportional gain
-    return k_p * error;
-  }
+// Function to calculate the control effort from the error.
+double LineFollower::calculateDistanceEffort(double error)
+{
+  double k_p = 0.1; // Proportional gain
```
Comment (ggupta9777) [inline src/line_follower.cpp]: Similar to naive implementation, this should be set from a function overriding the default value from constructor

File: test/test_line_follower_poor.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Test for goToGoal function
+TEST_F(TestNaiveLineFollower, TestgoToGoal)
+{  
+  // Define test ROS 2 Node
+  auto test_got_to_goal_node = std::make_shared<rclcpp::Node>("test_go_to_goal_node");
+  
+  // Define publisher to /tf topic
+  auto publisher_ = test_got_to_goal_node->create_publisher<tf2_msgs::msg::TFMessage>("tf", 10);
```
Comment (ggupta9777) [inline test/test_line_follower_poor.cpp]: This isn't a class member, so no underscore is needed as a suffix

File: test/test_line_follower_poor.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+  // Boolean to check if a velocity message is received on the bcr_bot/cmd_vel topic
+  bool received_twist=false;
```
Comment (ggupta9777) [inline test/test_line_follower_poor.cpp]: You can add a space on either side. Refer to cpp lint for the code in general

File: test/test_line_follower.cpp [hunk 10]
```diff
@@ -1,51 +1,251 @@
...
+    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscriber_;
+    geometry_msgs::msg::TransformStamped transform;
+    tf2_msgs::msg::TFMessage tf_message;
 };
 
-TEST_F(TestLineFollower, TestCalculateError) {
-  // Assuming you have a method in NaiveLineFollower to set waypoints for test purposes
-  node->setWaypoints(1.0, 0.0, 5.0, 0.0);  // Set waypoints for a horizontal line (y = 0)
+// Testing for identical waypoints to check for division by zero condition
+TEST_F(TestLineFollower, TestIdenticalWaypoints)
```
Comment (ggupta9777) [inline test/test_line_follower.cpp]: This test doesn't contribute meaningfully to the code so we should remove it

**Review by ggupta9777 — CHANGES_REQUESTED** [review]

**Comment by ggupta9777** [comment]
Screenshot from 2023-11-27 15-16-34

@topguns837 - can we remove these compile warnings?

**Comment by topguns837** [comment]
> Screenshot from 2023-11-27 15-16-34
> 
> @topguns837 - can we remove these compile warnings?

done

**Pull request merged**